File: c++/leetcode/1160. Find Words That Can Be Formed by Characters.cpp
class Solution
{
public:
  int countCharacters(vector<string> &words, string chars)
  {
    int len{};
    unordered_map<char, int> seen;

    for (auto i : chars)
      ++seen[i];

    for (auto i : words)
    {
      auto sn = seen;

      int k = 0;
      for (; k < i.size(); ++k)
      {
        --sn[i[k]];
        if (sn[i[k]] < 0)
          break;
      }

      if (k == i.size())
        len += i.size();
    }

    return len;
  }
};

File: javascript/leetcode/1160. Find Words That Can Be Formed by Characters.js
/**
 * @param {string[]} words
 * @param {string} chars
 * @return {number}
 */
var countCharacters = function (words, chars) {
  let len = 0;

  const seen = new Map();

  for (let i of chars) {
    if (seen.has(i)) seen.set(i, seen.get(i) + 1);
    else seen.set(i, 1);
  }

  for (let i of words) {
    const sn = new Map(seen.entries());

    let k = 0;
    for (let j of i) {
      if (sn.has(j)) sn.set(j, sn.get(j) - 1);
      else break;
      if (sn.get(j) < 0) break;
      ++k;
    }

    if (k === i.length) len += i.length;
  }

  return len;
};
